### PR TITLE
Stop CPU accounting to require CPU controller

### PR DESCRIPTION
**Pull request by ghost**
In the unified cgroup hierarchy, some CPU accounting information, `usage_usec`, `user_usec`, `system_usec` are always available in the `cpu.stat` file, while the CPU controller provides additional information `nr_periods`, `nr_throttled`, `throttled_usec` in the `cpu.stat` file. This behavior is defined in [https://www.kernel.org/doc/Documentation/cgroup-v2.txt](https://www.kernel.org/doc/Documentation/cgroup-v2.txt). On the other hand, systemd's `unit_get_cpu_usage_raw()` only uses `usage_usec`. Therefore, there is no need to enable the CPU controller in cgroup v2 when CPU accounting is required. This change prevents the CPU accounting from enabling the CPU controller when the unified cgroup is used, and meets the half of [RFE 9650](https://github.com/systemd/systemd/issues/9650).

### CODE REVIEW

**Comment by poettering** [comment]
Hmm, older kernels with cgroupsv2 didn't expose the cpu.stat file unless the CPU controller was enabled. On those we should really continue to enable the controller...

@htejun can you comment on this? What's the state there? Is it correct that on newer cgroupsv2 cpu.stat is always there regardless if "cpu" is enabled or not? Is there any nice way how we can detect kernels where it is like that?

**Comment by htejun** [comment]
Yeah, cpu.stat is always there regardless of cpu controller on newer kernels (v4.15+). On older kernels, cpu.stat requires cpu controller to be enabled, which depending on workloads could lead to noticeable overhead (we observe ~4% point higher system cpu time on context switch heavy workloads with 3-4 level of nesting).

**Comment by poettering** [comment]
@htejun hmm, is there a nice way to know whether it is there without "cpu" being enabled? i.e. i need to know if i have to enable the "cpu" controller for a cgroup, but i'd like to know that ahead of time. i mean, i could create a cgroup two levels deep and check if the file is there without enabling the controller, and then i know, but i'd prefer being able to do that without creating cgroups... i could also to a kernel version check, but we usually prefer checking for features rather then kernel versions... any suggestion?

**Comment by htejun** [comment]
Hmm... unfortunately, I can't think of a way to test it without creating a cgroup; however, two level deep shouldn't be necessary. If cpu controller isn't enabled at the root, checking whether the first level children have cpu.stat should be enough.

**Comment by ryutaroh-matsumoto** [comment]
With the unified cgroup hierarchy, systemd 239 on Linux 4.17.8 (debian) enables the CPU and IOcontrollers while systemd 239 on Linux 4.9.110 (debian) enables only IO controller, in addition to `memory pids` at `/sys/fs/cgroup/cgroup.controllers`. In both cases, `/sys/fs/cgroup` has three directories `system.slice`, `user.slice` and `init.scope` and we have `/sys/fs/cgroup/*/cpu.controllers == memory pids`. So I wonder why the CPU and the IO controllers are enabled at the root as the IO and the CPU seem unnecessary...

**Comment by keszybz** [comment]
I think a version check is acceptable in this case. In the unlikely scenario that somebody backports this feature into an older kernel, we will unnecessarily enable the controller. But this is just a small performance degradation without other problems. And creating cgroups just to check this is complicated and ugly and much more risky.

Please rework this to keep current behaviour if the kernel is older than 4.15. Something like this:
```C
assert_se(uname(&u) >= 0);
bool need_compat = str_verscmp(u.release, "4.15") < 0;
```

**Comment by ryutaroh-matsumoto** [comment]
Thank you for examining this issue. I will revise PR, hopefully on Saturday.

**Comment by ryutaroh-matsumoto** [comment]
I updated the commit to check the kernel version as instructed.

File: src/core/cgroup.c [hunk 1]
```diff
@@ -1062,15 +1063,30 @@ static void cgroup_context_apply(
...
+        bool need_cpumask_for_cpu_accounting;
+        struct utsname u;
+
+        /* In the unified cgroup hierarachy CPU accounting information is
+         * read from cpu.stat file. On Linux 4.15 and newer cpu.stat exists
+         * without CPU controller enabled. On the other hand, on older kernels
+         * CPU controller has to be enabled for cpu.stat to appear. Let's check it.
+         * Related discussion found at PR #9665. */
+        assert_se(uname(&u) >= 0);
+        need_cpumask_for_cpu_accounting = str_verscmp(u.release, "4.15") < 0;
```
Comment (poettering) [inline src/core/cgroup.c:1075]: hmm, doing this on each call might be needlessly expensive... could you turn the above two lines into a function of its own, that uses an internal `static` variable to cache the result? If you need inspiration how that function should look like, please look for the definition of urlify_enabled() in the sources (as one example), which does this correctly...

File: src/core/cgroup.c [hunk 2]
```diff
@@ -1062,15 +1063,30 @@ static void cgroup_context_apply(
...
-        if (c->cpu_accounting ||
-            cgroup_context_has_cpu_weight(c) ||
+        if (cgroup_context_has_cpu_weight(c) ||
             cgroup_context_has_cpu_shares(c) ||
             c->cpu_quota_per_sec_usec != USEC_INFINITY)
                 mask |= CGROUP_MASK_CPUACCT | CGROUP_MASK_CPU;
 
+        if (c->cpu_accounting) {
+                mask |= CGROUP_MASK_CPUACCT;
+                if(need_cpumask_for_cpu_accounting)
```
Comment (poettering) [inline src/core/cgroup.c:1086]: coding style nitpick: please add a space between `if` and the opening `(`

File: src/core/cgroup.c [hunk 3]
```diff
@@ -1062,15 +1063,30 @@ static void cgroup_context_apply(
...
+                        mask |= CGROUP_MASK_CPU;
```
Comment (poettering) [inline src/core/cgroup.c:1087]: hmm, on cgroupsv1 we should still add CPUACCT and CPU into the mask here as before, after all we always mount both controllers together there.

File: src/core/cgroup.c [hunk 4]
```diff
@@ -2476,9 +2492,6 @@ static int unit_get_cpu_usage_raw(Unit *u, nsec_t *ret) {
                 _cleanup_free_ char *val = NULL;
                 uint64_t us;
 
-                if ((u->cgroup_realized_mask & CGROUP_MASK_CPU) == 0)
-                        return -ENODATA;
-
```
Comment (poettering) [inline src/core/cgroup.c:2481]: i think this test should stay, but be bypassed if the version check holds (which is cheap if it becomes its own function as suggested above)

**Pull request closed without merging**